drivers/eeprom/eeprom_st25dv.c

```diff
@@ -0,0 +1,970 @@
+	ST25DV_REG_BLK_SIZE = 0x0016,
+	ST25DV_REG_IC_REF = 0x0017,
+	ST25DV_REG_UUID = 0x0018,
+	ST25DV_REG_IC_REV = 0x0020,
+	ST25DV_REG_I2C_PASSWD = 0x0900,
+	ST25DV_REG_I2C_SSO_DYN = 0x2004,
+} st25dv_reg_t;
+
+/** @brief Length of the MEM_SIZE field of the ST25DV */
+#define ST25DV_MEM_SIZE_LENGTH   2u
+/** @brief Length of the BLK_SIZE field of the ST25DV */
+#define ST25DV_BLK_SIZE_LENGTH   1u
+/** @brief Length of the IC_REV field of the ST25DV */
+#define ST25DV_IC_REV_LENGTH     1u
+/** @brief Length of ENDA registers */
+#define ST25DV_ENDA_LENGTH       1u
+/** @brief Length of I2C_PASSWD message */
+#define ST25DV_I2C_PASSWD_LENGTH 17u
+/** @brief Length of RFAxSS registers */
+#define ST25DV_RFASS_LENGTH      1u
+
+/**
+ * @brief I2C write timeout (ms)
+ *
+ * @details Max value: t_write + t_write * (max write bytes) / (internal page write)
+ * 5 + 5*(256/16)
+ */
+#define EEPROM_ST25DV_WRITE_TIMEOUT 85u
+
+/* I2C Security Session Open I2C_SSO_Dyn */
+#define EEPROM_ST25DV_I2C_SSO_DYN_I2CSSO_SHIFT 0
+#define EEPROM_ST25DV_I2C_SSO_DYN_I2CSSO_FIELD 0xFE
+#define EEPROM_ST25DV_I2C_SSO_DYN_I2CSSO_MASK  0x01
+
+/* RF X Access Protection RFAxSS */
+#define EEPROM_ST25DV_RFAXSS_PWD_CTRL_SHIFT 0
+#define EEPROM_ST25DV_RFAXSS_PWD_CTRL_FIELD 0xFC
+#define EEPROM_ST25DV_RFAXSS_PWD_CTRL_MASK  0x03
+#define EEPROM_ST25DV_RFAXSS_RW_PROT_SHIFT  2u
+#define EEPROM_ST25DV_RFAXSS_RW_PROT_FIELD  0xF3
+#define EEPROM_ST25DV_RFAXSS_RW_PROT_MASK   0x0C
+#define EEPROM_ST25DV_RFAXSS_ALL_SHIFT      0
+#define EEPROM_ST25DV_RFAXSS_ALL_MASK       0x0F
+
+/* I2C Access Protection I2CSS */
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A1_SHIFT 0
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A1_FIELD 0xFC
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A1_MASK  0x03
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A2_SHIFT 2u
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A2_FIELD 0xF3
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A2_MASK  0x0C
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A3_SHIFT 4u
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A3_FIELD 0xCF
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A3_MASK  0x30
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A4_SHIFT 6u
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A4_FIELD 0x3F
+#define EEPROM_ST25DV_I2CSS_RW_PROT_A4_MASK  0xC0
+
+struct eeprom_st25dv_config {
+	struct i2c_dt_spec bus;
+	uint16_t addr;
+	uint8_t power_state;
+	size_t size;
+};
+
+struct eeprom_st25dv_data {
+	struct k_mutex lock;
+};
+
+static int _st25dv_read(const struct eeprom_st25dv_config *config, uint16_t addr, off_t offset,
+			uint8_t *pdata, uint32_t count)
+{
+	int64_t timeout;
+	int ret = 0;
+	uint8_t wr_data[2] = {offset >> 8, offset & 0xff};
+	/*
+	 * A write cycle may be in progress so reads must be attempted
+	 * until the current write cycle should be completed.
+	 */
+	timeout = k_uptime_get() + EEPROM_ST25DV_WRITE_TIMEOUT;
+	while (1) {
+		int64_t now = k_uptime_get();
+
+		ret = i2c_write_read(config->bus.bus, addr, wr_data, 2, pdata, count);
+		if (!ret || now > timeout) {
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	}
```

Comment:
Have you seen the `WAIT_FOR()` macro in `include/zephyr/sys/util.h`?

Reply:
I did not know about the WAIT_FOR macro. I have updated the code to use it. The fact that I want to reuse the return value of the i2c function makes it a bit difficult to read but I guess it's not that harder to read than the manual loop I had earlier